articles/realtime.md

```diff
@@ -0,0 +1,259 @@
+
+Here are the some solutions to priority inversion:
+
+* Don't use locks (but sometimes, they are necessary)
+* Disable preemption for tasks holding locks (can lead to jitter)
+* Increase priority of task holding a lock
+* Use priority inheritance: a task that owns a lock inherits the priority of a task that tries to acquire the lock
+
+### Timing shots
+
+One real-time synchronization technique is when a thread calculates its next "shot" (the start of its next execution period).
+For example, if a thread is required to provide an update every 10 milliseconds, and it must complete an operation that takes 3-6 milliseconds, the thread should get the time before the operation, do the operation, and then wait for the remaining 7-4 milliseconds, based on the time measured after the operation.
+
+The most important consideration for the developer is to use a high precision timer, such as `nanosleep` on Linux platforms, while waiting.
+Otherwise the system will experience drift.
+
+### Spinlocks
+Spinlocks tend to cause clock drift.
+The developer should avoid implementing his own spinlocks.
+The RT Preempt patch replaces much of the kernel's spinlocks with mutexes, but this might not be guaranteed on all platforms.
+
+# Testing and Performance Benchmarking
+
+## cyclictest
+
+`cyclictest` is a simple Linux command line tool for measuring the jitter of a real-time environment.
+It takes as input a number of threads, a priority for the threads, and a scheduler type.
+It spins up `n` threads that sleep regular intervals (the sleep period can also be specified from the command line).
+
+For each thread, `cyclictest` measures the time between when the thread is supposed to wake up and when it actually wakes up.
+This statistic is the scheduling jitter in the system.
+If there are processes with non-deterministic blocking behavior running in the system, the jitter will grow to a large number (on the order of milliseconds), since the scheduler cannot meet the deadlines of the periodically sleeping threads profiled in the program.
+An ideal real-time system will have an average scheduling jitter on the order of nanoseconds or tens of nanoseconds.
+
+## Test Pipeline
+
+Instrumenting real-time code to validate its correctness may be tricky because the time overhead due to instrumentation.
+A simple benchmarking test that does not involve code instrumentation is as follows:
+
+On an RTLinux machine, start script that runs the process to be tested for some fixed time interval.
+Collect data with cyclictest to record the jitter in the system.
+A variety of command line tools could be used to add extra stress to the system in a controlled way, such as [cache calibrator](http://homepages.cwi.nl/~manegold/Calibrator/) to pollute the cache in an attempt to create pagefaults, or `fping`, which generates a large number of interrupts.
```

Comment:
The RTOS we have more experience with is Xenomai. It comes with its own [testing scripts](http://xenomai.org/2014/07/benchmarking-with-xeno-test/).

There's also the [stress](http://manpages.ubuntu.com/manpages/lucid/man1/stress.1.html) package for subjecting a system to different kinds of load.